### DISCUSSION TITLE
What steps to bring back Smooth Panning to MuseScore 4?

### DISCUSSION THREAD

**Discussion by davidlbangs** [musescore/MuseScore#18145] [General]
I am sticking with MuseScore 3, because I rely heavily on the Smooth Panning feature in MuseScore 3.
With MS3, I am able to turn on smoothPan options in Advanced Settings.  Then I can mute an instrument and play along with that part while hearing the other instruments.  At no point does the score lurch a whole screen at a time, and at all times I can see what is coming up.

In MuseScore 4, this feature is missing.  I've seen it reported, and every time someone reports it, people reply by linking to all the other reports.    This is not helpful because all the threads being linked are equally uninformative about what happened to this feature.  It does not seem to be assigned to anybody.

If nobody is working on it, I could try my hand at it.  I am retired as a senior developer at Microsoft, and am currently working on low key projects.   Before I jump in though, I'd like to know some background. 

1)  Is the core implementation or Smooth Panning still in the code?
2)  If the Smooth Panning options, or some improved version of these options, were added back into Advanced Settings, would the rest of the code respond to them? 
3)  Can someone who has significant experience with the code base give some suggestions about steps someone could take to reenable the same Smooth Panning functionality present in MuseScore 3. 

Thanks in advance for any replies. 
--David Bangs

**Comment by Jojo-Schmitz** [comment]
This is about #15212

**Comment by Jojo-Schmitz** [comment]
In 3.x there's `ScoreView::adjustCanvasPosition()` in scoreview.cpp and it has this bit of code:
```cpp
            // this code implements "continuous" panning
            // it could potentially be enabled via more panning options
            if (playBack && _cursor && seq->isPlaying() && preferences.getBool(PREF_PAN_SMOOTHLY_ENABLED)) {
                  // keep playback cursor pinned at 35% (or at the percent of controlCursorScreenPos + 5%)
                  xo = -curPosL * physicalZoomLevel() + marginLeft + width() * _panSettings.controlCursorScreenPos;
                  }
```

And this:
```cpp
void SmoothPanSettings::loadFromPreferences()
      {
      enabled = preferences.getBool(PREF_PAN_SMOOTHLY_ENABLED);
      controlModifierBase = preferences.getDouble(PREF_PAN_MODIFIER_BASE);
      if(mscore->currentScoreView() != nullptr)
            mscore->currentScoreView()->_controlModifier = controlModifierBase;
      controlModifierSteps = preferences.getDouble(PREF_PAN_MODIFIER_STEP);
      minContinuousModifier = preferences.getDouble(PREF_PAN_MODIFIER_MIN);
      maxContinuousModifier = preferences.getDouble(PREF_PAN_MODIFIER_MAX);
//      leftDistance = preferences.getDouble(PREF_PAN_DISTANCE_LEFT);
//      leftDistance1 = preferences.getDouble(PREF_PAN_DISTANCE_LEFT1);
//      leftDistance2 = preferences.getDouble(PREF_PAN_DISTANCE_LEFT2);
//      leftDistance3 = preferences.getDouble(PREF_PAN_DISTANCE_LEFT3);
//      leftMod1 = preferences.getDouble(PREF_PAN_MODIFIER_LEFT1);
//      leftMod2 = preferences.getDouble(PREF_PAN_MODIFIER_LEFT2);
//      leftMod3 = preferences.getDouble(PREF_PAN_MODIFIER_LEFT3);
//      rightDistance = preferences.getDouble(PREF_PAN_DISTANCE_RIGHT);
//      rightDistance1 = preferences.getDouble(PREF_PAN_DISTANCE_RIGHT1);
//      rightDistance2 = preferences.getDouble(PREF_PAN_DISTANCE_RIGHT2);
//      rightDistance3 = preferences.getDouble(PREF_PAN_DISTANCE_RIGHT3);
//      rightMod1 = preferences.getDouble(PREF_PAN_MODIFIER_RIGHT1);
//      rightMod2 = preferences.getDouble(PREF_PAN_MODIFIER_RIGHT2);
//      rightMod3 = preferences.getDouble(PREF_PAN_MODIFIER_RIGHT3);
//      normalWeight = preferences.getDouble(PREF_PAN_WEIGHT_NORMAL);
//      smartWeight = preferences.getDouble(PREF_PAN_WEIGHT_SMART);
//      advancedWeighting = preferences.getBool(PREF_PAN_WEIGHT_ADVANCED);
//      cursorTimerDuration = preferences.getInt(PREF_PAN_SMART_TIMER_DURATION);
      controlCursorScreenPos = preferences.getDouble(PREF_PAN_CURSOR_POS);
      teleportLeftEnabled = preferences.getBool(PREF_PAN_TELEPORT_LEFT);
      teleportRightEnabled = preferences.getBool(PREF_PAN_TELEPORT_RIGHT);
      }
```

**Comment by Jojo-Schmitz** [comment]
In Mu4 I find this:
```cpp
//! NOTE: Copied from ScoreView::adjustCanvasPosition
void NotationInteraction::showItem(const mu::engraving::EngravingItem* el, int staffIndex)
{
    if (!el) {
        return;
    }

    if (!configuration()->isAutomaticallyPanEnabled()) {
        return;
    }
```
I believe it is pretty save to state that the smooth scoll source code has been removed entirely

**Reply by davidlbangs** [reply]
Thank you.  I used git clone and got the project.  I found this reference in notationinteraction.cpp. 
Can I clone the MuseScore 3 code that you are looking at?

**Reply by Jojo-Schmitz** [reply]
It is in the same repo, in the 3.x branch

**Reply by davidlbangs** [reply]
Thank you.  I cloned the branch and see the code.  I'll take a break now, but plan to to study what is going on and see how to incorporate the changes.  However, I'm afraid people made a decision to remove this for a reason.  It might have been that all the options add a lot of clutter to settings and the developers did not understand the use case.  There are probably a lot of people who would have used this feature if the interace to turn it on weren't so hidden.  Personally, I think the majority of users would prefer smooth scrolling in continuous view.  Why play in continous view at all if you don't want to see the music scroll by correctly?

**Comment by MarcSabatella** [comment]
"Correctly" is subjective.  To me, the "correct" behavior is for the music to sit still on my piano until I reach my arm out and turn the page (or nod to my page turner to do the same); anything else is just an approximation :-).  Even when using continuous view in playback - some of us find it quite disorienting to see the music move while we're trying to read it.  Of course, seeing it suddenly jump is disorienting too, but at least that happens only at discrete times and not constantly.

So, I see the value of the smooth pan, but definitely would want to continue to have both available.

**Reply by davidlbangs** [reply]
Thank you.  I want to know all the perspectives.   For me, the problem is not being able to see what is next.  I don't personally get disoriented by looking at the center of the screen and watching the notes go by, but I do get disoriented by not seeing what is next.
